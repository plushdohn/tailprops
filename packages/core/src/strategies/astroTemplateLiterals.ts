import * as parser from "@babel/parser";
import generate from "@babel/generator";
import {
  escapeDollarsInString,
  getTailwindModifiersInAttribute,
} from "./utils/generic";
import {
  createClassAttributeFromRawTailprops,
  getHtmlTagsInsideTemplateLiterals,
  RawTailprop,
  removeSliceFromString,
  removeSlicesInString,
} from "./utils/templates";

export function transpileUsingAstroTemplateLiterals(
  source: string,
  options: {
    classAttributeKeyword: string;
    attributeFunctionId: string;
    generateSourceMaps?: boolean;
  } = {
    classAttributeKeyword: "class",
    attributeFunctionId: "$$addAttribute",
  }
) {
  const htmlTags = getHtmlTagsInsideTemplateLiterals(source);

  for (const tag of htmlTags.reverse()) {
    source =
      source.slice(0, tag.start) +
      transformHtmlTag(tag.content, options) +
      source.slice(tag.end);
  }

  const ast = parser.parse(source, { sourceType: "unambiguous" });

  const newSource = generate(ast, { sourceMaps: options.generateSourceMaps });

  return {
    code: newSource.code,
    map: newSource.map,
    ast,
  };
}

function transformHtmlTag(
  tag: string,
  options: {
    classAttributeKeyword: string;
    attributeFunctionId: string;
  }
) {
  const tailprops = findTailpropsInTag(tag, options.attributeFunctionId);

  tag = removeSlicesInString(tag, tailprops);

  const existingClassAttribute = findExistingClassAttributeInTag(tag, options);

  if (existingClassAttribute)
    tag = removeSliceFromString(tag, existingClassAttribute.slice);

  if (existingClassAttribute && existingClassAttribute.dynamic) {
    return addAttributeToTagUsingFunction(
      tag,
      {
        name: options.classAttributeKeyword,
        value: createClassAttributeFromRawTailprops(
          existingClassAttribute.slice.content,
          tailprops
        ),
      },
      options.attributeFunctionId
    );
  }

  return addStaticAttributeToTag(tag, {
    name: options.classAttributeKeyword,
    value: eval(
      createClassAttributeFromRawTailprops(
        existingClassAttribute ? existingClassAttribute.slice.content : null,
        tailprops
      )
    ),
  });
}

function findTailpropsInTag(
  tag: string,
  attributeFunctionIdentifier: string
): RawTailprop[] {
  const staticTailprops = findStaticTailpropsInTag(tag);
  const dynamicTailprops = findDynamicTailpropsInTag(
    tag,
    attributeFunctionIdentifier
  );

  if (dynamicTailprops.length > 0)
    throw new Error(
      "Expressions in Tailprops are not supported! Please use a flat string literal instead, or move the expression to the class attribute."
    );

  return staticTailprops;
}

function findStaticTailpropsInTag(tag: string): RawTailprop[] {
  const matches = [...tag.matchAll(/(tw(?:-\w*?)*?)=(".*?")/g)];

  return matches.map((m) => ({
    start: m.index!,
    end: m.index! + m[0].length,
    modifiers: getTailwindModifiersInAttribute(m[1]),
    content: m[2],
  }));
}

function findDynamicTailpropsInTag(
  tag: string,
  attributeFunctionId: string
): RawTailprop[] {
  const re = new RegExp(
    `\\\${${escapeDollarsInString(
      attributeFunctionId
    )}\\(([^,]*?),\\s?"(tw(?:-\\w*?)*?)"\\)}`,
    "g"
  );

  const matches = [...tag.matchAll(re)];

  return matches.map((m) => ({
    start: m.index!,
    end: m.index! + m[0].length,
    modifiers: getTailwindModifiersInAttribute(m[1]),
    content: m[2],
  }));
}

function findExistingClassAttributeInTag(
  tag: string,
  options: {
    attributeFunctionId: string;
    classAttributeKeyword: string;
  }
) {
  const dynamicClassAttribute = findExistingDynamicClassAttributeInTag(
    tag,
    options
  );

  if (dynamicClassAttribute)
    return {
      dynamic: true,
      slice: dynamicClassAttribute,
    };

  const staticClassAttribute = findExistingStaticClassAttributeInTag(tag);

  if (staticClassAttribute)
    return {
      dynamic: false,
      slice: staticClassAttribute,
    };

  return null;
}

function findExistingDynamicClassAttributeInTag(
  tag: string,
  options: {
    attributeFunctionId: string;
    classAttributeKeyword: string;
  }
) {
  const re = new RegExp(
    `\\\${${escapeDollarsInString(options.attributeFunctionId)}\\((.*?),\\s?"${
      options.classAttributeKeyword
    }"\\)}`
  );
  const match = tag.match(re);

  if (match === null) return null;

  return {
    start: match.index!,
    end: match.index! + match[0].length,
    content: match[1],
  };
}

function findExistingStaticClassAttributeInTag(tag: string) {
  const match = tag.match(/class=("[^"]*?")/);

  if (match === null) return null;

  return {
    start: match.index!,
    end: match.index! + match[0].length,
    content: match[1],
  };
}

function addStaticAttributeToTag(
  tag: string,
  attribute: { name: string; value: string }
) {
  const index = tag.indexOf(">");

  tag =
    tag.slice(0, index) +
    ` ${attribute.name}="${attribute.value}"` +
    tag.slice(index);

  return tag;
}

function addAttributeToTagUsingFunction(
  tag: string,
  attribute: { name: string; value: string },
  functionId: string
) {
  const index = tag.indexOf(">");

  tag =
    tag.slice(0, index) +
    `\${${functionId}(${attribute.value},"${attribute.name}")}` +
    tag.slice(index);

  return tag;
}
