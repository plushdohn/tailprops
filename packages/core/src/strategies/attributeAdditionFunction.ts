/**
 * This strategy exists mainly to address Svelte's hydration phase.
 *
 * Changing the attributes on the server-rendered component is not
 * enough for the changes to persist as they get rewritten by the
 * client-side hydration.
 *
 * As of now, Svelte creates some object with a function property
 * which is called during hydration. This function is responsible
 * for setting the attributes on the element.
 *
 * This strategy will transform the function block, removing all
 * calls to "tw" attributes and transform them into a single
 * "class" attribute with the modifiers applied.
 *
 * The original code looks something like this:
 *
 * m: function hydrate() {
 *  attr_dev(h1, "class", "existing-class");
 *  attr_dev(h1, "tw", "bg-red-500");
 *  attr_dev(h1, "tw-hover", "bg-red-400");
 * }
 *
 * And it will be transformed into:
 *
 * m: function hydrate() {
 *  attr_dev(h1, "class", "existing-class" + " " + "bg-red-500" + " " + "hover:bg-red-400");
 * }
 */

import generate from "@babel/generator";
import * as parser from "@babel/parser";
import traverse, { NodePath } from "@babel/traverse";
import * as t from "@babel/types";
import {
  getTailwindModifiersInAttribute,
  getTailwindPropertiesInString,
  joinPropertiesUsingModifiers,
} from "./utils/generic";

type TailpropNodesMap = {
  [key: string]: string[];
};

type ClassNodesMap = {
  [key: string]: {
    path: NodePath<t.CallExpression>;
    expression: t.Expression;
  };
};

export function transpileUsingAttributeAdditionFunction(
  source: string,
  options: {
    ast?: any;
    generateSourceMaps?: boolean;
    attributeFunctionId: string;
    classAttributeKeyword: string;
  } = {
    classAttributeKeyword: "class",
    attributeFunctionId: "attr_dev",
  }
) {
  let ast = options.ast;

  if (!ast) {
    ast = parser.parse(source, { sourceType: "unambiguous" });
  }

  transformFunctionBlock(ast, options);

  const newSource = generate(ast, { sourceMaps: options.generateSourceMaps });

  return {
    code: newSource.code,
    map: newSource.map,
    ast,
  };
}

function transformFunctionBlock(
  ast: parser.ParseResult<t.File>,
  options: {
    attributeFunctionId: string;
    classAttributeKeyword: string;
  }
) {
  const tailpropNodes: TailpropNodesMap = {};
  const classNodes: ClassNodesMap = {};

  traverseOnFunctionExpression(ast, (functionPath) => {
    subTraverseOnCallExpression(
      functionPath,
      options.attributeFunctionId,
      (callPath) => {
        const { node } = callPath;

        if (node.arguments[0].type !== "Identifier") {
          console.log(
            "Unexpected non-identifier in attribute addition function:",
            JSON.stringify(node.arguments[0])
          );
          return;
        }

        const tailprop = tryGetCallExpressionAsTailprop(
          callPath,
          options.attributeFunctionId
        );

        if (tailprop !== null) {
          const currentProps = tailpropNodes[tailprop.elementName] || [];

          tailpropNodes[tailprop.elementName] = [
            ...currentProps,
            tailprop.prop,
          ];

          return callPath.remove();
        }

        const classAttribute = tryGetCallExpressionAsClassAttribute(
          callPath,
          options.attributeFunctionId
        );

        if (classAttribute) {
          classNodes[classAttribute.elementName] = {
            path: callPath,
            expression: callPath.node.arguments[2] as t.Expression,
          };
        }
      }
    );

    /**
     * Re-traverse function body to eliminate all stored class
     * nodes, since we'll create a new one.
     * TODO: Find a way to avoid doing this. It's probably best
     * to not remove them altogether.
     */
    subTraverseOnCallExpression(
      functionPath,
      options.attributeFunctionId,
      (callPath) => {
        const classAttribute = tryGetCallExpressionAsClassAttribute(
          callPath,
          options.attributeFunctionId
        );

        if (classAttribute && tailpropNodes[classAttribute.elementName]) {
          classAttribute.path.remove();
        }
      }
    );

    // Add new class node for each tailprop node
    for (const [element, tailprops] of Object.entries(tailpropNodes)) {
      addCallExpressionToFunctionBody(
        functionPath.node,
        t.callExpression(t.identifier(options.attributeFunctionId), [
          t.identifier(element),
          t.stringLiteral(options.classAttributeKeyword),
          mergeExpressionsWithSpaces([
            ...(classNodes[element] ? [classNodes[element].expression] : []),
            ...tailprops.map((p) => t.stringLiteral(p)),
          ]),
        ])
      );
    }
  });
}

function tryGetCallExpressionAsTailprop(
  path: NodePath<t.CallExpression>,
  attributeFunctionId: string
) {
  if (
    path.node.callee.type === "Identifier" &&
    path.node.callee.name === attributeFunctionId &&
    path.node.arguments[0].type === "Identifier" &&
    path.node.arguments[1].type === "StringLiteral" &&
    path.node.arguments[1].value.startsWith("tw")
  ) {
    if (path.node.arguments[2].type !== "StringLiteral")
      throw new Error(
        "Expressions in Tailprops are not supported! Please use a flat string literal or move the expression to the class attribute."
      );

    const modifiers = getTailwindModifiersInAttribute(
      path.node.arguments[1].value
    );
    const properties = getTailwindPropertiesInString(
      path.node.arguments[2].value
    );

    return {
      elementName: path.node.arguments[0].name,
      prop: joinPropertiesUsingModifiers(properties, modifiers),
    };
  }

  return null;
}

function tryGetCallExpressionAsClassAttribute(
  path: NodePath<t.CallExpression>,
  attributeFunctionId: string
) {
  if (
    path.node.callee.type === "Identifier" &&
    path.node.callee.name === attributeFunctionId &&
    path.node.arguments[0].type === "Identifier" &&
    path.node.arguments[1].type === "StringLiteral" &&
    path.node.arguments[1].value === "class"
  ) {
    return {
      elementName: path.node.arguments[0].name,
      path,
      expression: path.node.arguments[2] as t.Expression,
    };
  }

  return null;
}

function mergeExpressionsWithSpaces(expressions: t.Expression[]) {
  return expressions.slice(1).reduce((acc, expression) => {
    return t.binaryExpression(
      "+",
      t.binaryExpression("+", acc, t.stringLiteral(" ")),
      expression
    );
  }, expressions[0]);
}

function subTraverseOnCallExpression(
  path: NodePath<t.FunctionExpression | t.ObjectMethod>,
  functionIdentifier: string,
  callback: (path: NodePath<t.CallExpression>) => void
) {
  traverse(
    path.node,
    {
      CallExpression(path) {
        if (
          path.node.callee.type === "Identifier" &&
          path.node.callee.name === functionIdentifier
        ) {
          callback(path);
        }
      },
    },
    path.scope,
    path.state,
    path.parentPath
  );
}

function addCallExpressionToFunctionBody(
  node: t.FunctionExpression | t.ObjectMethod,
  callExpression: t.CallExpression
) {
  node.body.body.push(t.expressionStatement(callExpression));
}

function traverseOnFunctionExpression(
  ast: parser.ParseResult<t.File>,
  callback: (path: NodePath<t.FunctionExpression | t.ObjectMethod>) => void
) {
  traverse(ast, {
    FunctionExpression(path) {
      if (path.node.id) callback(path);
    },
    ObjectMethod(path) {
      if (path.node.key.type === "Identifier") callback(path);
    },
  });
}
