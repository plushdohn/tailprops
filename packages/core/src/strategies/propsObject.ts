import generate from "@babel/generator";
import * as parser from "@babel/parser";
import traverse, { NodePath } from "@babel/traverse";
import * as t from "@babel/types";
import {
  getTailwindModifiersInAttribute,
  getTailwindPropertiesInString,
  joinPropertiesUsingModifiers,
} from "./utils";

export function transpileJsUsingPropsObject(
  source: string,
  options: {
    classAttribute: string;
    generateSourceMaps?: boolean;
    ast?: parser.ParseResult<t.File>;
  }
) {
  let ast = options.ast;

  if (!ast) {
    ast = parser.parse(source, { sourceType: "unambiguous" });
  }

  traverse(ast, {
    ObjectExpression(path) {
      const tailwindClasses = extractTailpropsClassesFromObjectNode(path, "tw");

      if (tailwindClasses) {
        for (const expression of tailwindClasses) {
          appendExpressionToObjectProperty(
            path.node,
            options.classAttribute,
            expression
          );
        }
      }
    },
  });

  const newSource = generate(ast, { sourceMaps: options.generateSourceMaps });

  return {
    code: newSource.code,
    map: newSource.map,
  };
}

function extractTailpropsClassesFromObjectNode(
  path: NodePath<t.ObjectExpression>,
  prefix: string
) {
  let expressions: t.ObjectProperty["value"][] = [];

  for (const [i, p] of path.node.properties.entries()) {
    const property = getObjectPropertyAsTailprop(p, prefix);

    if (property !== null) {
      delete path.node.properties[i];

      const modifiers = getTailwindModifiersInAttribute(property.key);

      if (property.expression.type === "StringLiteral") {
        const properties = getTailwindPropertiesInString(
          property.expression.value
        );

        property.expression.value = joinPropertiesUsingModifiers(
          properties,
          modifiers
        );

        expressions.push(property.expression);

        continue;
      }

      traverse(
        property.expression,
        {
          StringLiteral({ node }) {
            const properties = getTailwindPropertiesInString(node.value);

            node.value = joinPropertiesUsingModifiers(properties, modifiers);
          },
        },
        path.scope,
        null,
        path.parentPath
      );

      expressions.push(property.expression);
    }
  }

  // Clear up dangling nodes
  path.node.properties = path.node.properties.filter((p) => p !== undefined);

  return expressions;
}

function getObjectPropertyAsTailprop(
  p: t.ObjectProperty | t.SpreadElement | t.ObjectMethod,
  prefix: string
) {
  if (p.type === "ObjectProperty") {
    if (p.key.type === "Identifier" && p.key.name.startsWith(prefix)) {
      return {
        key: p.key.name,
        expression: p.value,
      };
    }

    if (p.key.type === "StringLiteral" && p.key.value.startsWith(prefix)) {
      return {
        key: p.key.value,
        expression: p.value,
      };
    }
  }

  return null;
}

function appendExpressionToObjectProperty(
  node: t.ObjectExpression,
  key: string,
  value: t.ObjectProperty["value"]
) {
  // Find property with matching key
  const property = node.properties.find((p) => {
    if ("key" in p) {
      if (p.key.type === "Identifier" && p.key.name === key) {
        return true;
      }

      if (p.key.type === "StringLiteral" && p.key.value === key) {
        return true;
      }
    }
  }) as t.ObjectProperty | undefined;

  if (!property) {
    node.properties.push(t.objectProperty(t.identifier(key), value));
    return;
  }

  if (!isObjectPropertyConcatenable(property.value)) {
    throw new Error(
      `Value type for property ${key} can't be concatenated on: ${property.value.type}`
    );
  }

  if (!isObjectPropertyConcatenable(value)) {
    throw new Error(
      `Passed a value type for key ${key} that can't be used for concatenation: ${property.value.type}`
    );
  }

  property.value = t.binaryExpression(
    "+",
    property.value as t.Expression,
    t.binaryExpression("+", t.stringLiteral(" "), value as t.Expression)
  );
}

function isObjectPropertyConcatenable(expression: t.ObjectProperty["value"]) {
  return (
    expression.type !== "RestElement" &&
    expression.type !== "NullLiteral" &&
    expression.type !== "AssignmentPattern" &&
    expression.type !== "ArrayPattern" &&
    expression.type !== "ObjectPattern"
  );
}
