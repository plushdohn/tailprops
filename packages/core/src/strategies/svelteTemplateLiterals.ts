import * as parser from "@babel/parser";
import generate from "@babel/generator";
import { getTailwindModifiersInAttribute } from "./utils/generic";
import {
  createClassAttributeFromRawTailprops,
  getHtmlTagsInsideTemplateLiterals,
  RawTailprop,
  removeSliceFromString,
  removeSlicesInString,
} from "./utils/templates";

export function transpileUsingSvelteTemplateLiterals(
  source: string,
  options: {
    classAttributeKeyword: string;
    attributeFunctionId: string;
    generateSourceMaps?: boolean;
  } = {
    classAttributeKeyword: "class",
    attributeFunctionId: "add_attribute",
  }
) {
  const htmlTags = getHtmlTagsInsideTemplateLiterals(source);

  for (const tag of htmlTags.reverse()) {
    source =
      source.slice(0, tag.start) +
      transformHtmlTag(tag.content, options) +
      source.slice(tag.end);
  }

  const ast = parser.parse(source, { sourceType: "unambiguous" });

  const newSource = generate(ast, { sourceMaps: options.generateSourceMaps });

  return {
    code: newSource.code,
    map: newSource.map,
    ast,
  };
}

function transformHtmlTag(
  tag: string,
  options: {
    classAttributeKeyword: string;
    attributeFunctionId: string;
  }
) {
  const tailprops = findTailpropsInTag(tag, options.attributeFunctionId);

  tag = removeSlicesInString(tag, tailprops);

  const existingClassAttribute = findExistingClassAttributeInTag(tag, options);

  if (existingClassAttribute)
    tag = removeSliceFromString(tag, existingClassAttribute);

  tag = addAttributeToTagUsingFunction(
    tag,
    {
      name: options.classAttributeKeyword,
      value: createClassAttributeFromRawTailprops(
        existingClassAttribute ? existingClassAttribute.content : null,
        tailprops
      ),
    },
    options.attributeFunctionId
  );

  return tag;
}

function findTailpropsInTag(
  tag: string,
  attributeFunctionIdentifier: string
): RawTailprop[] {
  const staticTailprops = findStaticTailpropsInTag(tag);
  const dynamicTailprops = findDynamicTailpropsInTag(
    tag,
    attributeFunctionIdentifier
  );

  if (dynamicTailprops.length > 0)
    throw new Error(
      "Expressions in Tailprops are not supported! Please use a flat string literal instead, or move the expression to the class attribute."
    );

  return staticTailprops;
}

function findStaticTailpropsInTag(tag: string): RawTailprop[] {
  const matches = [...tag.matchAll(/(tw(?:-\w*?)*?)="\${(.*?)\}"/g)];

  return matches.map((m) => ({
    start: m.index!,
    end: m.index! + m[0].length,
    modifiers: getTailwindModifiersInAttribute(m[1]),
    content: m[2],
  }));
}

function findDynamicTailpropsInTag(
  tag: string,
  attributeFunctionId: string
): RawTailprop[] {
  const re = new RegExp(
    `\\\${${attributeFunctionId}\\("(tw(?:-\\w*?)*?)",\\s?(.*?),.*?\\)}`,
    "g"
  );

  const matches = [...tag.matchAll(re)];

  return matches.map((m) => ({
    start: m.index!,
    end: m.index! + m[0].length,
    modifiers: getTailwindModifiersInAttribute(m[1]),
    content: m[2],
  }));
}

function findExistingClassAttributeInTag(
  tag: string,
  options: {
    attributeFunctionId: string;
    classAttributeKeyword: string;
  }
) {
  const dynamicClassAttribute = findExistingDynamicClassAttributeInTag(
    tag,
    options
  );

  if (dynamicClassAttribute) return dynamicClassAttribute;

  const staticClassAttribute = findExistingStaticClassAttributeInTag(tag);

  if (staticClassAttribute) return staticClassAttribute;

  return null;
}

function findExistingDynamicClassAttributeInTag(
  tag: string,
  options: {
    attributeFunctionId: string;
    classAttributeKeyword: string;
  }
) {
  const match = tag.match(
    new RegExp(
      `\\\${${options.attributeFunctionId}\\("${options.classAttributeKeyword}",(.*?),.*?\\)}`
    )
  );

  if (match === null) return null;

  return {
    start: match.index!,
    end: match.index! + match[0].length,
    content: match[1],
  };
}

function findExistingStaticClassAttributeInTag(tag: string) {
  const match = tag.match(/class="\${(.*?)}"/);

  if (match === null) return null;

  return {
    start: match.index!,
    end: match.index! + match[0].length,
    content: match[1],
  };
}

function addAttributeToTagUsingFunction(
  tag: string,
  attribute: { name: string; value: string },
  functionId: string
) {
  const index = tag.indexOf(">");

  tag =
    tag.slice(0, index) +
    `\${${functionId}("${attribute.name}",${attribute.value}, 0)}` + // i have no idea what "0" does here tbh
    tag.slice(index);

  return tag;
}
