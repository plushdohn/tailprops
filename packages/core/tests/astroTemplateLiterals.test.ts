import { transpileUsingAstroTemplateLiterals } from "../src/strategies/astroTemplateLiterals";

describe("template literals strategy", () => {
  it("concatenates basic strings correctly", () => {
    const source =
      'console.log("hello world");\nfunction Test() { return `<html><h1 tw="bg-red-500" tw-hover="bg-red-400">ciao</h1></html>`} ';

    const result = transpileUsingAstroTemplateLiterals(source);

    expect(result.code).toMatchSnapshot();
  });

  it("concatenates to existing classes", () => {
    const source =
      'function Test() { return `<html><h1 class="test-class" tw-hover="bg-red-400">ciao</h1></html>`}';

    const result = transpileUsingAstroTemplateLiterals(source);

    expect(result.code).toMatchSnapshot();
  });

  it("handles complicated class expressions with different order", () => {
    const source =
      'function Test() { return `<html><h1${$$addAttribute(someState ? "test-class" : "another-class", "class")} tw="flex" tw-2xl="flex-row">ciao</h1></html>`}';

    const result = transpileUsingAstroTemplateLiterals(source);

    expect(result.code).toMatchSnapshot();
  });

  it("should throw on expressions", () => {
    const source =
      'function Test() { return `<html><h1${$$addAttribute("test", "tw-hover")}>ciao</h1></html>`}';

    expect(() => transpileUsingAstroTemplateLiterals(source)).toThrow(
      /expressions/i
    );
  });
});
