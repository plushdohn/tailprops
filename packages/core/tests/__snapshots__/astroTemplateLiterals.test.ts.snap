// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template literals strategy concatenates basic strings correctly 1`] = `
"console.log("hello world");
function Test() {
  return \`<html><h1   class="bg-red-500 hover:bg-red-400">ciao</h1></html>\`;
}"
`;

exports[`template literals strategy concatenates to existing classes 1`] = `
"function Test() {
  return \`<html><h1   class="test-class hover:bg-red-400">ciao</h1></html>\`;
}"
`;

exports[`template literals strategy handles complicated class expressions with different order 1`] = `
"function Test() {
  return \`<html><h1  \${$$addAttribute((someState ? "test-class" : "another-class") + " " + "flex" + " " + "2xl:flex-row", "class")}>ciao</h1></html>\`;
}"
`;
